calico/felix/ipsets.py

```diff
@@ -380,15 +390,26 @@ def _remove_mapping(self, tag_id, profile_id, endpoint_id, ip_address):
         :param EndpointId endpoint_id: ID of the endpoint
         :param str ip_address: IP address to remove
         """
-        ep_ids = self.ip_owners_by_tag[tag_id][ip_address][profile_id]
-        ep_ids.discard(endpoint_id)
-        if not ep_ids:
-            del self.ip_owners_by_tag[tag_id][ip_address][profile_id]
-            if not self.ip_owners_by_tag[tag_id][ip_address]:
-                del self.ip_owners_by_tag[tag_id][ip_address]
-                self._dirty_tags.add(tag_id)
+        owners = self.ip_owners_by_tag[tag_id][ip_address]
+        removed_mapping = (profile_id, endpoint_id)
+        if owners == removed_mapping:
+            # This was the sole owner of the IP in the tag, remove it.
+            _log.debug("%s was sole owner of IP %s, IP no longer in tag",
+                       removed_mapping, ip_address)
+            del self.ip_owners_by_tag[tag_id][ip_address]
             if not self.ip_owners_by_tag[tag_id]:
+                _log.debug("Tag %s now empty, removing", tag_id)
                 del self.ip_owners_by_tag[tag_id]
+            self._dirty_tags.add(tag_id)
+        elif isinstance(owners, set):
+            assert len(owners) != 1, ("ip_owners_by_tag entry should never "
+                                      "be a set with 1 entry")
+            _log.debug("Tag %s still contains IP %s", tag_id, ip_address)
+            owners.discard(removed_mapping)
+            if len(owners) == 1:
+                _log.debug("IP %s now only has one owner, replacing set with "
+                           "single tuple", ip_address)
+                self.ip_owners_by_tag[tag_id][ip_address], = owners
```

Comment:
:+1: for the above; I actually think this would be much more readable as

    self.ip_owners_by_tag[tag_id][ip_address] = owners.pop()